File: benchmarks/successor/representation-v1/transformed/deterministic/csharp/gold/006-transition-validation/Program.cs
using System.Text.Json;

namespace OrderFlow;

public sealed class Customer
{
    public string? Id { get; init; }
    public string? Tier { get; init; }
}

public sealed class Order
{
    public string Id { get; init; } = "";
    public DateTimeOffset CreatedAt { get; init; }
    public string Status { get; init; } = "";
    public int? Priority { get; init; }
    public DateTimeOffset? DueAt { get; init; }
    public Customer? Customer { get; init; }
}

public sealed class Request
{
    public string Operation { get; init; } = "";
    public Order?[]? Orders { get; init; }
    public DateTimeOffset? AsOf { get; init; }
    public string? Id { get; init; }
    public string? ToStatus { get; init; }
}

public sealed record Response(string[] Ids);

public sealed record TransitionResponse(string Id, string Status);

public static class Program
{
    private static readonly JsonSerializerOptions Options = new()
    {
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    public static void Main()
    {
        string? loc_ey5euhrnbl;
        while ((loc_ey5euhrnbl = Console.ReadLine()) is not null)
        {
            try
            {
                var loc_zk64hdzcrs = JsonSerializer.Deserialize<Request>(loc_ey5euhrnbl, Options)
                    ?? throw new InvalidOperationException("Request was null");
                var loc_r43bavyf4z = Handle(loc_zk64hdzcrs);
                Console.WriteLine(JsonSerializer.Serialize(loc_r43bavyf4z, Options));
            }
            catch (Exception loc_fnvfrnyz4a)
            {
                Console.WriteLine(JsonSerializer.Serialize(new { error = loc_fnvfrnyz4a.Message }, Options));
            }
        }
    }

    private static object Handle(Request loc_zk64hdzcrs)
    {
        if (IsStatus(loc_zk64hdzcrs.Operation, "transition"))
        {
            return Transition(loc_zk64hdzcrs);
        }

        var orders = NormalizeOrders(loc_zk64hdzcrs.Orders);
        IEnumerable<Order> selected;

        if (IsStatus(loc_zk64hdzcrs.Operation, "ready"))
        {
            selected = Ready(orders);
        }
        else if (IsStatus(loc_zk64hdzcrs.Operation, "overdue"))
        {
            var asOf = loc_zk64hdzcrs.AsOf
                ?? throw new InvalidOperationException("asOf is required for overdue");
            selected = Overdue(orders, asOf);
        }
        else if (IsStatus(loc_zk64hdzcrs.Operation, "atRisk"))
        {
            var asOf = loc_zk64hdzcrs.AsOf
                ?? throw new InvalidOperationException("asOf is required for atRisk");
            selected = AtRisk(orders, asOf);
        }
        else if (IsStatus(loc_zk64hdzcrs.Operation, "vipReady"))
        {
            selected = Ready(orders).Where(IsVip);
        }
        else
        {
            throw new InvalidOperationException($"Unknown operation: {loc_zk64hdzcrs.Operation}");
        }

        return new Response(selected.Select(loc_dagq2dkqyn => loc_dagq2dkqyn.Id).ToArray());
    }

    private static TransitionResponse Transition(Request loc_zk64hdzcrs)
    {
        if (string.IsNullOrEmpty(loc_zk64hdzcrs.Id))
        {
            throw new InvalidOperationException("id is required for transition");
        }

        if (string.IsNullOrEmpty(loc_zk64hdzcrs.ToStatus))
        {
            throw new InvalidOperationException("toStatus is required for transition");
        }

        var matches = NormalizeOrders(loc_zk64hdzcrs.Orders)
            .Where(loc_dagq2dkqyn =>
                string.Equals(loc_dagq2dkqyn.Id, loc_zk64hdzcrs.Id, StringComparison.Ordinal))
            .Take(2)
            .ToArray();

        if (matches.Length == 0)
        {
            throw new InvalidOperationException("order not found for transition");
        }

        if (matches.Length > 1)
        {
            throw new InvalidOperationException("duplicate order id for transition");
        }

        var target = CanonicalTarget(matches[0].Status, loc_zk64hdzcrs.ToStatus);
        return new TransitionResponse(matches[0].Id, target);
    }

    private static string CanonicalTarget(string? source, string target)
    {
        if (IsStatus(source, "pending"))
        {
            if (IsStatus(target, "processing"))
            {
                return "processing";
            }

            if (IsStatus(target, "cancelled"))
            {
                return "cancelled";
            }
        }
        else if (IsStatus(source, "processing"))
        {
            if (IsStatus(target, "completed"))
            {
                return "completed";
            }

            if (IsStatus(target, "cancelled"))
            {
                return "cancelled";
            }
        }

        throw new InvalidOperationException("invalid transition");
    }

    private static Order[] NormalizeOrders(Order?[]? orders) =>
        orders?.OfType<Order>().ToArray() ?? [];

    private static IEnumerable<Order> Ready(IEnumerable<Order> orders) =>
        orders
            .Where(loc_dagq2dkqyn => IsStatus(loc_dagq2dkqyn.Status, "pending"))
            .OrderByDescending(loc_dagq2dkqyn => loc_dagq2dkqyn.Priority ?? 0)
            .ThenBy(loc_dagq2dkqyn => loc_dagq2dkqyn.CreatedAt)
            .ThenBy(loc_dagq2dkqyn => loc_dagq2dkqyn.Id, StringComparer.Ordinal);

    private static IEnumerable<Order> Overdue(
        IEnumerable<Order> orders,
        DateTimeOffset asOf) =>
        orders
            .Where(loc_dagq2dkqyn =>
                IsActive(loc_dagq2dkqyn.Status)
                && loc_dagq2dkqyn.DueAt.HasValue
                && loc_dagq2dkqyn.DueAt.Value < asOf)
            .OrderBy(loc_dagq2dkqyn => loc_dagq2dkqyn.DueAt!.Value)
            .ThenByDescending(loc_dagq2dkqyn => loc_dagq2dkqyn.Priority ?? 0)
            .ThenBy(loc_dagq2dkqyn => loc_dagq2dkqyn.Id, StringComparer.Ordinal);

    private static IEnumerable<Order> AtRisk(
        IEnumerable<Order> orders,
        DateTimeOffset asOf)
    {
        var upperBound = asOf.AddHours(24);

        return orders
            .Where(loc_dagq2dkqyn =>
                IsActive(loc_dagq2dkqyn.Status)
                && loc_dagq2dkqyn.DueAt.HasValue
                && loc_dagq2dkqyn.DueAt.Value >= asOf
                && loc_dagq2dkqyn.DueAt.Value < upperBound)
            .OrderBy(loc_dagq2dkqyn => loc_dagq2dkqyn.DueAt!.Value)
            .ThenByDescending(loc_dagq2dkqyn => loc_dagq2dkqyn.Priority ?? 0)
            .ThenBy(loc_dagq2dkqyn => loc_dagq2dkqyn.Id, StringComparer.Ordinal);
    }

    private static bool IsVip(Order loc_dagq2dkqyn) =>
        loc_dagq2dkqyn.Customer is not null
        && (IsStatus(loc_dagq2dkqyn.Customer.Tier, "gold")
            || IsStatus(loc_dagq2dkqyn.Customer.Tier, "platinum"));

    private static bool IsActive(string? status) =>
        IsStatus(status, "pending") || IsStatus(status, "processing");

    private static bool IsStatus(string? actual, string expected) =>
        string.Equals(actual, expected, StringComparison.OrdinalIgnoreCase);
}

File: benchmarks/successor/representation-v1/transformed/deterministic/fsharp/gold/006-transition-validation/Program.fs
open System
open System.Text.Json

[<CLIMutable>]
type Customer =
    { id: string
      tier: string }

[<CLIMutable>]
type Order =
    { id: string
      createdAt: DateTimeOffset
      status: string
      priority: Nullable<int>
      dueAt: Nullable<DateTimeOffset>
      customer: Customer }

[<CLIMutable>]
type Request =
    { operation: string
      orders: Order array
      asOf: Nullable<DateTimeOffset>
      id: string
      toStatus: string }

[<CLIMutable>]
type Response = { ids: string array }

[<CLIMutable>]
type TransitionResponse =
    { id: string
      status: string }

let options =
    JsonSerializerOptions(
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    )

let isStatus actual expected =
    String.Equals(actual, expected, StringComparison.OrdinalIgnoreCase)

let isActive status =
    isStatus status "pending" || isStatus status "processing"

let priorityOf (loc_dagq2dkqyn: Order) =
    if loc_dagq2dkqyn.priority.HasValue then loc_dagq2dkqyn.priority.Value else 0

let normalizeOrders (orders: Order array) =
    if Object.ReferenceEquals(orders, null) then
        [||]
    else
        orders
        |> Array.choose (fun loc_dagq2dkqyn ->
            if isNull (box loc_dagq2dkqyn) then None else Some loc_dagq2dkqyn)

let compareReady (left: Order) (right: Order) =
    let byPriority = compare (priorityOf right) (priorityOf left)

    if byPriority <> 0 then
        byPriority
    else
        let byCreatedAt = DateTimeOffset.Compare(left.createdAt, right.createdAt)

        if byCreatedAt <> 0 then
            byCreatedAt
        else
            StringComparer.Ordinal.Compare(left.id, right.id)

let compareDuePriorityId (left: Order) (right: Order) =
    let byDueAt = DateTimeOffset.Compare(left.dueAt.Value, right.dueAt.Value)

    if byDueAt <> 0 then
        byDueAt
    else
        let byPriority = compare (priorityOf right) (priorityOf left)

        if byPriority <> 0 then
            byPriority
        else
            StringComparer.Ordinal.Compare(left.id, right.id)

let ready (orders: Order array) =
    orders
    |> Array.filter (fun loc_dagq2dkqyn -> isStatus loc_dagq2dkqyn.status "pending")
    |> Array.sortWith compareReady

let overdue asOf (orders: Order array) =
    orders
    |> Array.filter (fun loc_dagq2dkqyn ->
        isActive loc_dagq2dkqyn.status
        && loc_dagq2dkqyn.dueAt.HasValue
        && loc_dagq2dkqyn.dueAt.Value < asOf)
    |> Array.sortWith compareDuePriorityId

let atRisk (asOf: DateTimeOffset) (orders: Order array) =
    let upperBound = asOf.AddHours(24.0)

    orders
    |> Array.filter (fun loc_dagq2dkqyn ->
        isActive loc_dagq2dkqyn.status
        && loc_dagq2dkqyn.dueAt.HasValue
        && loc_dagq2dkqyn.dueAt.Value >= asOf
        && loc_dagq2dkqyn.dueAt.Value < upperBound)
    |> Array.sortWith compareDuePriorityId

let isVip (loc_dagq2dkqyn: Order) =
    not (isNull (box loc_dagq2dkqyn.customer))
    && (isStatus loc_dagq2dkqyn.customer.tier "gold"
        || isStatus loc_dagq2dkqyn.customer.tier "platinum")

let canonicalTarget source target =
    if isStatus source "pending" then
        if isStatus target "processing" then
            "processing"
        elif isStatus target "cancelled" then
            "cancelled"
        else
            raise (InvalidOperationException("invalid transition"))
    elif isStatus source "processing" then
        if isStatus target "completed" then
            "completed"
        elif isStatus target "cancelled" then
            "cancelled"
        else
            raise (InvalidOperationException("invalid transition"))
    else
        raise (InvalidOperationException("invalid transition"))

let transition (loc_zk64hdzcrs: Request) : TransitionResponse =
    if String.IsNullOrEmpty(loc_zk64hdzcrs.id) then
        raise (InvalidOperationException("id is required for transition"))

    if String.IsNullOrEmpty(loc_zk64hdzcrs.toStatus) then
        raise (InvalidOperationException("toStatus is required for transition"))

    let matches =
        loc_zk64hdzcrs.orders
        |> normalizeOrders
        |> Array.filter (fun loc_dagq2dkqyn ->
            String.Equals(loc_dagq2dkqyn.id, loc_zk64hdzcrs.id, StringComparison.Ordinal))

    if matches.Length = 0 then
        raise (InvalidOperationException("order not found for transition"))

    if matches.Length > 1 then
        raise (InvalidOperationException("duplicate order id for transition"))

    { id = matches[0].id
      status = canonicalTarget matches[0].status loc_zk64hdzcrs.toStatus }

let query (loc_zk64hdzcrs: Request) : Response =
    let orders = normalizeOrders loc_zk64hdzcrs.orders

    let selected =
        match loc_zk64hdzcrs.operation with
        | operation when isStatus operation "ready" ->
            ready orders
        | operation when isStatus operation "overdue" ->
            if not loc_zk64hdzcrs.asOf.HasValue then
                raise (InvalidOperationException("asOf is required for overdue"))

            overdue loc_zk64hdzcrs.asOf.Value orders
        | operation when isStatus operation "atRisk" ->
            if not loc_zk64hdzcrs.asOf.HasValue then
                raise (InvalidOperationException("asOf is required for atRisk"))

            atRisk loc_zk64hdzcrs.asOf.Value orders
        | operation when isStatus operation "vipReady" ->
            orders
            |> ready
            |> Array.filter isVip
        | operation ->
            raise (InvalidOperationException($"Unknown operation: {operation}"))

    selected
    |> Array.map (fun loc_dagq2dkqyn -> loc_dagq2dkqyn.id)
    |> fun ids -> { ids = ids }

let handle (loc_zk64hdzcrs: Request) : obj | null =
    if isStatus loc_zk64hdzcrs.operation "transition" then
        box (transition loc_zk64hdzcrs)
    else
        box (query loc_zk64hdzcrs)

let mutable running = true

while running do
    match Console.ReadLine() with
    | null -> running <- false
    | loc_ey5euhrnbl ->
        try
            let loc_zk64hdzcrs = JsonSerializer.Deserialize<Request>(loc_ey5euhrnbl, options)

            match loc_zk64hdzcrs with
            | null -> raise (InvalidOperationException("Request was null"))
            | loc_zk64hdzcrs ->
                let loc_r43bavyf4z = handle loc_zk64hdzcrs
                Console.WriteLine(JsonSerializer.Serialize(loc_r43bavyf4z, options))
        with loc_fnvfrnyz4a ->
            Console.WriteLine(JsonSerializer.Serialize({| error = loc_fnvfrnyz4a.Message |}, options))
